### PR TITLE
Fix stack offset in cfg with Istackoffset

### PR DESCRIPTION
**Pull request by gretay-js**
Cfg to linear needs to take into account Istackoffset instructions when calculating virtual stack adjustment.

- change Ladjust_trap_depth in traps to Ladjust_stack_offset in bytes
- change trap_depth to stack_offset in Cfg

This PR is work in progress, mostly because cfgize needs to be fixed and more testing is needed with FDO. There is some debug printing that needs to be in a separate PR.

### CODE REVIEW

**Comment by gretay-js** [comment]
This is ready for review. 

I had to revert #186 because the information can no longer be derived from other fields.
I changed `cfgize` to compute traps for dead handlers, if there are any, but not sure if the extra passes are ever triggered, perhaps we should leave that change out.
I also cleaned up the PR, removing unrelated changes such as printing improvements and a  fix for unifying trap stacks, which I'll submit separately.

**Comment by gretay-js** [comment]
I've updated the PR after an offline discussion with @xclerc:
- no need to revert #186, compute `can_raise_interproc` from `can_raise` and `exn` fields instead.
- leave dead blocks in cfgizes with stack_offset_invalid

File: backend/cfg/cfgize.ml [hunk 1]
```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
...
-    | Op _ | Call _ | Reloadretaddr | Prologue ->
-      ( record_handler stack exceptional_successor
-          ~can_raise:(Cfg.can_raise_basic instr.desc),
-        instr )
+      stack_offset, traps, term
+
+  let rec update_block :
+      Cfg.t -> Label.t -> stack_offset:int -> traps:handler_stack -> unit =
+   fun cfg label ~stack_offset ~traps ->
+    let process_basic :
```
Comment (xclerc) [inline backend/cfg/cfgize.ml:607]: (For consistency, could be moved at
the same level as `process_terminator`.)

File: backend/cfg/cfgize.ml [hunk 2]
```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
...
-   fun exceptional_successor stack term ->
-    let term = Cfg.set_trap_depth term (succ (List.length stack)) in
+      int * handler_stack * Cfg.terminator Cfg.instruction =
+   fun ~stack_offset ~traps term ->
+    assert (term.stack_offset = invalid_stack_offset);
+    let term = check_and_set_stack_offset term ~stack_offset ~traps in
     match term.desc with
+    | Tailcall (Self _) when List.length traps <> 0 || stack_offset <> 0 ->
+      Misc.fatal_error
+        "Cfgize.Trap_depth_and_exn.terminator: unexpected handler on self \
```
Comment (xclerc) [inline backend/cfg/cfgize.ml:595]:
```suggestion
        "Cfgize.Trap_depth_and_exn.preocess_terminator: unexpected handler on self \
```

File: backend/cfg/cfgize.ml [hunk 3]
```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
...
+      let instr = check_and_set_stack_offset instr ~stack_offset ~traps in
+      match instr.desc with
+      | Pushtrap { lbl_handler } ->
+        update_block cfg lbl_handler ~stack_offset ~traps;
+        stack_offset, lbl_handler :: traps, instr
+      | Poptrap -> begin
+        match traps with
+        | [] ->
+          Misc.fatal_error
+            "Cfgize.Trap_depth_and_exn.basic: trying to pop from an empty stack"
```
Comment (xclerc) [inline backend/cfg/cfgize.ml:622]:
```suggestion
            "Cfgize.Trap_depth_and_exn.process_basic: trying to pop from an empty stack"
```

File: backend/cfg/cfgize.ml [hunk 4]
```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
...
     let block = Cfg.get_block_exn cfg label in
     let was_invalid =
-      if block.trap_depth = invalid_trap_depth
+      if block.stack_offset = invalid_stack_offset
       then true
       else begin
-        assert (block.trap_depth = succ (List.length stack));
+        assert (
+          block.stack_offset = compute_stack_offset ~stack_offset ~traps
+          && block.stack_offset >= Proc.trap_size_in_bytes * List.length traps);
```
Comment (xclerc) [inline backend/cfg/cfgize.ml:643]: Why is it `>=`?

File: backend/cfg/cfgize.ml [hunk 5]
```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
...
           block.exn <- Some handler_label;
-          update_block cfg handler_label handler_stack)
-        exceptional_successor
+          update_block cfg handler_label ~stack_offset ~traps:rest
+      end
     end
 
   let update_cfg : Cfg.t -> unit =
-   fun cfg -> update_block cfg cfg.entry_label []
+   fun cfg ->
```
Comment (xclerc) [inline backend/cfg/cfgize.ml:680]: I think the main change is that the module
no longer relies on `can_raise` over basic
and terminator instructions. We should maybe
leave a comment to this effect. Indeed, we
will want to revisit the behavior as soon as
we decide it is not taboo to remove a dead
exception handler.

**Review by xclerc — APPROVED** [review]
Looks correct, but some parts are subtle;
the more testing the better!

(I have a feeling some conditions such as
`trap_depth < 0` may have been too lax,
but they are gone after this pull request.)

**Comment by xclerc** [comment]
> * leave dead blocks in cfgizes with stack_offset_invalid

I feel compelled to add that:
- exception handlers are in this version never dead;
- dead blocks will be removed by a latter pass.

The combination means that no block with an invalid
offset will remain in the CFG. This is key, as having a
block will an invalid offset is likely to create problems
(read segfaults) down the line when we switch to a
CFG pipeline.

File: backend/cfg/linear_to_cfg.ml [hunk 6]
```diff
@@ -220,12 +218,13 @@ let check_traps t label (block : C.basic_block) =
       T.print trap_stack_at_start);
     match T.to_list_exn trap_stack_at_start with
     | trap_labels ->
-      if List.compare_length_with trap_labels block.trap_depth <> 0
+      if (List.length trap_labels - 1) * Proc.trap_size_in_bytes
+         > block.stack_offset
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/cfg/linear_to_cfg.ml:222]: The invariant is ```block.stack_offset = x + y``` where `x` is "trap stack depth in bytes" and `y` is "offset derived from Istackoffset operations".

The confusing thing is that the argument `stack_offset` in `Cfgize.Trap_depth_and_exn`  functions has a different meaning from the field `block.stack_offset`. The argument `stack_offset` means only `y`, as evident from `Cfgize.Trap_depth_and_exn.compute_stack_offset`. Perhaps it should be renamed.

[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/cfg/linear_to_cfg.ml:222]: Both x and y are non-negative.

[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/linear_to_cfg.ml:222]: Maybe, or add a comment (basically copy-pasting your reply).

File: backend/cfg/cfgize.ml [hunk 7]
```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
 module Trap_depth_and_exn = struct
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/cfg/cfgize.ml]: done

File: backend/cfg/cfgize.ml (hunk 2 above)
Comment (gretay-js) [reply backend/cfg/cfgize.ml]: fixed.

File: backend/cfg/cfgize.ml (hunk 1 above)
Comment (gretay-js) [reply backend/cfg/cfgize.ml]: fixed

File: backend/cfg/cfgize.ml (hunk 3 above)
Comment (gretay-js) [reply backend/cfg/cfgize.ml]: fixed.

File: backend/cfg/cfgize.ml (hunk 4 above)
Comment (gretay-js) [reply backend/cfg/cfgize.ml:653]: See the answer to a similar question in `linear_to_cfg`, except in cfgize `traps` starts from empty, i.e., has no dummy interproc handler, so no need for `-1`.

File: backend/cfg/cfgize.ml (hunk 5 above)
Comment (gretay-js) [reply backend/cfg/cfgize.ml:690]: I added a comment at the top the this module.

File: backend/cfg/linear_to_cfg.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/cfg/linear_to_cfg.ml:222]: Added a comment in `Cfgize.Stack_offset_and_exn`.

**Review by mshinwell — APPROVED** [review]
I've only looked at the changes owned by me.

**Pull request merged**